# Exploring 471 records from 1543 (Kickback from Lex's R Code)
***
***DAEN690***

***George Mason University***

***Author:*** Grace Cox (Team LEGO)

***Date:*** November 15, 2021
***
`This Jupyter Notebook explores 471 records (of the 1543 OTHER records) that were returned after being ran through Lex's R Code. The process for exploring these records is as follows:`

1. Explore the 'DATASET' field, which indicates what type of Origin Location Reference appears within the 'REMARKS' field for each record (e.g., AIRPORT, NAVAID, AIRPORT DIRECTLY OVER, NAVAID DIRECTLY OVER, RUNWAY, FIX RADIAL, HELIPORT, etc) 
<br>

2. For each of the unique Origin Location References that appear in the 'DATASET' field:

    a. Run the uas_loc_parts() function of the dataframe 
    
    b. Link to the appropriate Origin Location Reference linking function (e.g., link_airportsCleaned_ident(), link_airportsCleaned_icao(), nav_link())
        i. Separate the field that were successfully linked with the Origin Location Reference Dataset(s) and those that were not. Those that were not linked successfully will be moved to the exceptions file for further review.
    c. Run the successfully linked dataframe through the uas_lat_long() function to calculate the UAS Latitude/Longitude coordinates for each respective record/remark. 

***
### INPUT: 
*Dataframe (read from a .csv file) containing 471 records, of 1543 records that did not hit on the initial 'XX NM dir XXX' regular expressions, and was not initially picked up by our Runways regular expressions.*

### OUTPUT:
<font color='red'>**EXCPETIONS FILE**</font>

*Designated Point: 1 record (no link to Designated Points)*

*FAA Airport Identifier: 18 records (14 do not link to Airports; 4 have no distance information)*

*Reporting Facility used as Identifier: 95 records (no link to Airports or NAVAID Systems)*

<font color='green'>**UAS LATITUDE/LONGITUDE COORDINATE CALCULATED**</font>

*FAA Airport Identifier: 347 records*

*NAVAID System: 10 records*




## Import Statements

In [1]:
# Import Statements
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import chart_studio.plotly as py
import re
import geopy
from geopy.distance import geodesic

from IPython.display import display, HTML

In [2]:
output471 = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/Grace_All_Points_Standard_471OUTOF1543.csv')

In [3]:
#output471

In [4]:
output471_df = output471[['REMARKS', 'DATASET', 'UASLOCATION']]
output471_df

REMARKS           DATASET  \
0    Aircraft observed a large uas 2-5 NM away at h...            NAVAID   
1    Aircraft reported an orange quad copter uas 3 ...            NAVAID   
2    Aircraft reported an orange quad copter uas mo...            NAVAID   
3    Aircraft observed a uas while S bound at 2,000...  DESIGNATED POINT   
4    Aircraft observed a 20-30 pound uas pass withi...           AIRPORT   
..                                                 ...               ...   
466  Aircraft observed a uas  while northbound at 2...      FIXED RADIAL   
467  Aircraft reported a yellow drone approximately...      FIXED RADIAL   
468  Aircraft reported a small yellow drone DCA0660...      FIXED RADIAL   
469  DCA-2: Aircraft reported a blue and red uas at...      FIXED RADIAL   
470  Aircraft reported a white quad copter at the B...      FIXED RADIAL   

           UASLOCATION  
0          21NM SW ANX  
1          26 NM S LTJ  
2         18NM ESE TNV  
3      2 NM N of TUGGZ  
4           1NM SE NEW  
..                 ...  
466  012 NM 030 of AGC  
467  012 NM 036 of DCA  
468  030 NM 066 of DCA  
469  008 NM 201 of DCA  
470  010 NM 320 of BRV  

[471 rows x 3 columns]

In [5]:
# Determine Unique Origin Location References within the 471 record kickback
output471_df['DATASET'].unique()

array(['NAVAID', 'DESIGNATED POINT', 'AIRPORT', 'AIRPORT(?)',
       'FIXED RADIAL'], dtype=object)

# AIRPORTS

In [6]:
output471_airport = output471_df[output471_df['DATASET'] == 'AIRPORT'].reset_index()
output471_airport

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
240    244  Aircraft reported a white uas off the left sid...  AIRPORT   
241    245  Aircraft reported a YAS at 4,500 feet while W ...  AIRPORT   
242    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
243    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
244    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION  
0       1NM SE NEW  
1      2 NM SW BJC  
2    10 NM NNW VNY  
3        3NM N LIT  
4     4.5 NM W SLC  
..             ...  
240  3 NM W of DPA  
241   15 NM SE PDZ  
242   7 NM ENE DAL  
243   20 NM SE JFK  
244     5 NM W DFW  

[245 rows x 4 columns]

In [7]:
# Export to .csv 
output471_airport.to_csv('airport245.csv', index = False)

In [8]:
def uas_loc_parts(dataframe):
    '''
    This function takes the 'UASLOCATION' column from Lex's 'Kickbacks' and breaks it into the different parts of the 
    UAS Location, to include: Distance (NM), Direction Abbreviation, Airport Identifier
    
    input : dataframe is a df that includes the 'UASLOCATION' field
    output : the dataframe provided as input to the user with additional fields to include: 'Distance_NM', 'Bearing', 
                                                                                            'IDENT'
    '''
    uasLoc_col = dataframe['UASLOCATION']
    dist_list = []
    bearing_list = []
    ident_list = []
    dist_kilo = []
    bearDegree = []
    
    # Dictionary of Bearing Abbreviation to Degrees Conversions
    bearing_deg = {
    'N' : 0,
    'NNE' : 23,
    'NE' : 45,
    'ENE' : 68,
    'E' : 90,
    'ESE' : 113,
    'SE' : 135,
    'SSE' : 158,
    'S' : 180,
    'SSW' : 203,
    'SW' : 225,
    'WSW' : 248,
    'W' : 270,
    'WNW' : 293,
    'NW' : 315,
    'NNW' : 338
    }
    
    beardeg_list = list(bearing_deg.keys())
    
    for i in range(len(dataframe)):
        # Get Distance (NM) from UAS LOCATION and Convert Distance to Kilometers
        distNM = re.sub("[^\d\.]","", str(re.findall(r"[-+]?\d*\.\d+|\d+", uasLoc_col[i].split()[0])))
        if distNM == '':
            dist_list.append(0)
            dist_kilo.append(0.00)
        else:
            dist_list.append(distNM)
            dist_kilo.append(float(distNM) * 1.852)
        
        # Get IDENT from UASLOCATION
        ident = uasLoc_col[i].split()[-1].replace('of', '')
        ident_list.append(ident)
        
        # Get Bearing from UASLOCATION and Convert Bearing to Degrees
        #bearing = uasLoc_col[i].replace('NM', '').replace('of','').split()[-2]
        bearing = re.sub('(NM|of|OF)', ' ', uasLoc_col[i]).split()[-2]
        bearing_list.append(bearing)
        if bearing.isdigit():
            bearDegree.append(bearing)
        else:
            bearDegree.append(bearing_deg.get(bearing))
        
    # Append new fields to dataframe provided as input    
    dataframe['Distance_NM'] = dist_list
    dataframe['Distance_Kilometers'] = dist_kilo
    dataframe['Bearing'] = bearing_list
    dataframe['Bearing_Degrees'] = bearDegree
    dataframe['IDENT'] = ident_list
    
    
    return dataframe

In [9]:
test = uas_loc_parts(output471_airport)
test

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
240    244  Aircraft reported a white uas off the left sid...  AIRPORT   
241    245  Aircraft reported a YAS at 4,500 feet while W ...  AIRPORT   
242    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
243    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
244    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0       1NM SE NEW           1                1.852      SE              135   
1      2 NM SW BJC           2                3.704      SW              225   
2    10 NM NNW VNY          10               18.520     NNW              338   
3        3NM N LIT           3                5.556       N                0   
4     4.5 NM W SLC         4.5                8.334       W              270   
..             ...         ...                  ...     ...              ...   
240  3 NM W of DPA           3                5.556       W              270   
241   15 NM SE PDZ          15               27.780      SE              135   
242   7 NM ENE DAL           7               12.964     ENE               68   
243   20 NM SE JFK          20               37.040      SE              135   
244     5 NM W DFW           5                9.260       W              270   

    IDENT  
0     NEW  
1     BJC  
2     VNY  
3     LIT  
4     SLC  
..    ...  
240   DPA  
241   PDZ  
242   DAL  
243   JFK  
244   DFW  

[245 rows x 9 columns]

In [10]:
# Export to .csv 
test.to_csv('test245.csv', index = False)

In [11]:
def link_airportsCleaned(dataframe):
    '''
    This function links the provided dataframe to the airports_cleaned.csv file.
    
    input : dataframe is a df with an 'IDENT' field on which we will link to the airports_cleaned.csv
    output : the dataframe provided as input with airport lat/long fields appended
    '''
    
    # Read in Cleaned Airports dataset
    airportsC = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/airports_cleaned.csv')
    
    return pd.merge(dataframe, airportsC, on='IDENT', how='left')

In [12]:
test1 = link_airportsCleaned(test)
test1

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
240    244  Aircraft reported a white uas off the left sid...  AIRPORT   
241    245  Aircraft reported a YAS at 4,500 feet while W ...  AIRPORT   
242    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
243    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
244    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0       1NM SE NEW           1                1.852      SE              135   
1      2 NM SW BJC           2                3.704      SW              225   
2    10 NM NNW VNY          10               18.520     NNW              338   
3        3NM N LIT           3                5.556       N                0   
4     4.5 NM W SLC         4.5                8.334       W              270   
..             ...         ...                  ...     ...              ...   
240  3 NM W of DPA           3                5.556       W              270   
241   15 NM SE PDZ          15               27.780      SE              135   
242   7 NM ENE DAL           7               12.964     ENE               68   
243   20 NM SE JFK          20               37.040      SE              135   
244     5 NM W DFW           5                9.260       W              270   

    IDENT                             GLOBAL_ID  \
0     NEW  10BA3990-72A1-4F14-89DB-FC7DF04B54F7   
1     BJC  338AC1EC-7F12-4F3D-932E-E37882849344   
2     VNY  F0AFB4E6-592A-46B2-A76A-5376CC5A0016   
3     LIT  8FBCF8C3-B29D-470D-B4AF-93E8218EDFD5   
4     SLC  CB41ECB2-ADC7-49A5-9C05-81854AFD415C   
..    ...                                   ...   
240   DPA  2FD31A12-EF68-4339-891A-D84B47A5EE7B   
241   PDZ                                   NaN   
242   DAL  48203599-A2E2-467A-BB88-DAC28F3A7FA6   
243   JFK  DD5FB154-0787-4B0F-8554-C033FAFA0155   
244   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   

                                       NAME   LATITUDE   LONGITUDE ICAO_ID  
0                                 Lakefront  30.042417  -90.028250    KNEW  
1                      Rocky Mountain Metro  39.908806 -105.117194    KBJC  
2                                  Van Nuys  34.209806 -118.489972    KVNY  
3    Bill and Hillary Clinton Ntl/Adams Fld  34.729444  -92.224778    KLIT  
4                       Salt Lake City Intl  40.788388 -111.977773    KSLC  
..                                      ...        ...         ...     ...  
240                                  Dupage  41.907053  -88.247995    KDPA  
241                                     NaN        NaN         NaN     NaN  
242                         Dallas Love Fld  32.845945  -96.850877    KDAL  
243                     John F Kennedy Intl  40.639928  -73.778693    KJFK  
244                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  

[245 rows x 14 columns]

In [13]:
test1.dropna()

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
239    243  Aircraft reported a white fixed wing uas at 2,...  AIRPORT   
240    244  Aircraft reported a white uas off the left sid...  AIRPORT   
242    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
243    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
244    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0       1NM SE NEW           1                1.852      SE              135   
1      2 NM SW BJC           2                3.704      SW              225   
2    10 NM NNW VNY          10               18.520     NNW              338   
3        3NM N LIT           3                5.556       N                0   
4     4.5 NM W SLC         4.5                8.334       W              270   
..             ...         ...                  ...     ...              ...   
239   10 NM SW DFW          10               18.520      SW              225   
240  3 NM W of DPA           3                5.556       W              270   
242   7 NM ENE DAL           7               12.964     ENE               68   
243   20 NM SE JFK          20               37.040      SE              135   
244     5 NM W DFW           5                9.260       W              270   

    IDENT                             GLOBAL_ID  \
0     NEW  10BA3990-72A1-4F14-89DB-FC7DF04B54F7   
1     BJC  338AC1EC-7F12-4F3D-932E-E37882849344   
2     VNY  F0AFB4E6-592A-46B2-A76A-5376CC5A0016   
3     LIT  8FBCF8C3-B29D-470D-B4AF-93E8218EDFD5   
4     SLC  CB41ECB2-ADC7-49A5-9C05-81854AFD415C   
..    ...                                   ...   
239   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   
240   DPA  2FD31A12-EF68-4339-891A-D84B47A5EE7B   
242   DAL  48203599-A2E2-467A-BB88-DAC28F3A7FA6   
243   JFK  DD5FB154-0787-4B0F-8554-C033FAFA0155   
244   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   

                                       NAME   LATITUDE   LONGITUDE ICAO_ID  
0                                 Lakefront  30.042417  -90.028250    KNEW  
1                      Rocky Mountain Metro  39.908806 -105.117194    KBJC  
2                                  Van Nuys  34.209806 -118.489972    KVNY  
3    Bill and Hillary Clinton Ntl/Adams Fld  34.729444  -92.224778    KLIT  
4                       Salt Lake City Intl  40.788388 -111.977773    KSLC  
..                                      ...        ...         ...     ...  
239                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  
240                                  Dupage  41.907053  -88.247995    KDPA  
242                         Dallas Love Fld  32.845945  -96.850877    KDAL  
243                     John F Kennedy Intl  40.639928  -73.778693    KJFK  
244                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  

[231 rows x 14 columns]

In [14]:
noMatch_airports = test1[test1['LATITUDE'].isnull()]
len(noMatch_airports)

14

In [15]:
# Export to .csv 
noMatch_airports.to_csv('airports245_nomatch14.csv', index = False)

In [16]:
true_airports = test1[test1['LATITUDE'].notnull()].reset_index(drop=True)
true_airports

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
226    243  Aircraft reported a white fixed wing uas at 2,...  AIRPORT   
227    244  Aircraft reported a white uas off the left sid...  AIRPORT   
228    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
229    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
230    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0       1NM SE NEW           1                1.852      SE              135   
1      2 NM SW BJC           2                3.704      SW              225   
2    10 NM NNW VNY          10               18.520     NNW              338   
3        3NM N LIT           3                5.556       N                0   
4     4.5 NM W SLC         4.5                8.334       W              270   
..             ...         ...                  ...     ...              ...   
226   10 NM SW DFW          10               18.520      SW              225   
227  3 NM W of DPA           3                5.556       W              270   
228   7 NM ENE DAL           7               12.964     ENE               68   
229   20 NM SE JFK          20               37.040      SE              135   
230     5 NM W DFW           5                9.260       W              270   

    IDENT                             GLOBAL_ID  \
0     NEW  10BA3990-72A1-4F14-89DB-FC7DF04B54F7   
1     BJC  338AC1EC-7F12-4F3D-932E-E37882849344   
2     VNY  F0AFB4E6-592A-46B2-A76A-5376CC5A0016   
3     LIT  8FBCF8C3-B29D-470D-B4AF-93E8218EDFD5   
4     SLC  CB41ECB2-ADC7-49A5-9C05-81854AFD415C   
..    ...                                   ...   
226   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   
227   DPA  2FD31A12-EF68-4339-891A-D84B47A5EE7B   
228   DAL  48203599-A2E2-467A-BB88-DAC28F3A7FA6   
229   JFK  DD5FB154-0787-4B0F-8554-C033FAFA0155   
230   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   

                                       NAME   LATITUDE   LONGITUDE ICAO_ID  
0                                 Lakefront  30.042417  -90.028250    KNEW  
1                      Rocky Mountain Metro  39.908806 -105.117194    KBJC  
2                                  Van Nuys  34.209806 -118.489972    KVNY  
3    Bill and Hillary Clinton Ntl/Adams Fld  34.729444  -92.224778    KLIT  
4                       Salt Lake City Intl  40.788388 -111.977773    KSLC  
..                                      ...        ...         ...     ...  
226                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  
227                                  Dupage  41.907053  -88.247995    KDPA  
228                         Dallas Love Fld  32.845945  -96.850877    KDAL  
229                     John F Kennedy Intl  40.639928  -73.778693    KJFK  
230                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  

[231 rows x 14 columns]

`4 remarks that are missing valid distance information regarding the UAS location`

In [56]:
airports_invalidDist = true_airports[true_airports['Distance_NM'] == 0].reset_index(drop=True)
airports_invalidDist

index                                            REMARKS  DATASET  \
0     24  Aircraft observed a uas 100 feet high off the ...  AIRPORT   
1     28  Aircraft observed a uas off the left side whil...  AIRPORT   
2    126  Aircraft reported a uas 2NM NM E of PAE whileS...  AIRPORT   
3    139  Aircraft reported a silver uas off the left si...  AIRPORT   

     UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0   NM SW of APA           0                  0.0      SW              225   
1    NM S of PTW           0                  0.0       S              180   
2    NM E of PAE           0                  0.0       E               90   
3   NM NW of SFO           0                  0.0      NW              315   

  IDENT                             GLOBAL_ID                          NAME  \
0   APA  F45128FD-465B-4CD2-9CCA-F022A4152561                    Centennial   
1   PTW  164605B4-1F5A-4259-99DF-F1E81793C5E0                  Heritage Fld   
2   PAE  32EB01A7-F6A3-4C0F-8238-D92BB8F2B019  Snohomish County (Paine Fld)   
3   SFO  AF8BC34D-2D11-4CDB-9E87-33A0D430C055            San Francisco Intl   

    LATITUDE   LONGITUDE ICAO_ID  
0  39.570119 -104.849293    KAPA  
1  40.239556  -75.556722    KPTW  
2  47.907317 -122.282094    KPAE  
3  37.618806 -122.375417    KSFO

In [57]:
# Export to .csv 
airports_invalidDist.to_csv('airNoDist04OF471_EXCEPTIONS.csv', index = False)

`227 records with valid distance information regarding UAS location`

In [17]:
airports_validDist = true_airports[true_airports['Distance_NM'] != 0].reset_index(drop=True)
airports_validDist

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
222    243  Aircraft reported a white fixed wing uas at 2,...  AIRPORT   
223    244  Aircraft reported a white uas off the left sid...  AIRPORT   
224    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
225    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
226    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0       1NM SE NEW           1                1.852      SE              135   
1      2 NM SW BJC           2                3.704      SW              225   
2    10 NM NNW VNY          10               18.520     NNW              338   
3        3NM N LIT           3                5.556       N                0   
4     4.5 NM W SLC         4.5                8.334       W              270   
..             ...         ...                  ...     ...              ...   
222   10 NM SW DFW          10               18.520      SW              225   
223  3 NM W of DPA           3                5.556       W              270   
224   7 NM ENE DAL           7               12.964     ENE               68   
225   20 NM SE JFK          20               37.040      SE              135   
226     5 NM W DFW           5                9.260       W              270   

    IDENT                             GLOBAL_ID  \
0     NEW  10BA3990-72A1-4F14-89DB-FC7DF04B54F7   
1     BJC  338AC1EC-7F12-4F3D-932E-E37882849344   
2     VNY  F0AFB4E6-592A-46B2-A76A-5376CC5A0016   
3     LIT  8FBCF8C3-B29D-470D-B4AF-93E8218EDFD5   
4     SLC  CB41ECB2-ADC7-49A5-9C05-81854AFD415C   
..    ...                                   ...   
222   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   
223   DPA  2FD31A12-EF68-4339-891A-D84B47A5EE7B   
224   DAL  48203599-A2E2-467A-BB88-DAC28F3A7FA6   
225   JFK  DD5FB154-0787-4B0F-8554-C033FAFA0155   
226   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   

                                       NAME   LATITUDE   LONGITUDE ICAO_ID  
0                                 Lakefront  30.042417  -90.028250    KNEW  
1                      Rocky Mountain Metro  39.908806 -105.117194    KBJC  
2                                  Van Nuys  34.209806 -118.489972    KVNY  
3    Bill and Hillary Clinton Ntl/Adams Fld  34.729444  -92.224778    KLIT  
4                       Salt Lake City Intl  40.788388 -111.977773    KSLC  
..                                      ...        ...         ...     ...  
222                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  
223                                  Dupage  41.907053  -88.247995    KDPA  
224                         Dallas Love Fld  32.845945  -96.850877    KDAL  
225                     John F Kennedy Intl  40.639928  -73.778693    KJFK  
226                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW  

[227 rows x 14 columns]

In [18]:
# Export to .csv 
airports_validDist.to_csv('airports245_validDist227.csv', index = False)

In [19]:
def uas_lat_long(dataframe):  
    '''
    This function takes in a .csv file (string) of SkyWatch records for which the Remarks are in the Standard Format
    and uses the bearing_dist_originIdent() function to output the dataframe containing complete UAS Location information
    in terms of the UAS' Latitude and Longitude.
    
    @param dataframe : The dataframe that includes the fields 'LATITUDE', 'LONGITUDE', 'Bearing_Degrees', 
                        'Distance_Kilometers' to have UAS Lat/Long values calculated for
    
    @output new_df: returns a dataframe containing the Remark, UAS Location, FAA Airport Identifier, Distance (in NM) from 
                    the airport, and bearing abbreviation (WHERE ALL BEARINGS ARE VALID)
                
                *** it should be noted that the full dataset that includes remarks for which UAS latitudes and longitudes
                    were calculated is located in the StandardRemark_UAS_LatLong_validBearing_FAAairIdent.csv file.
                    
    
    '''
    uas_lat = []
    uas_long = []
    
    for i in range(len(dataframe)):
        lat_airport = pd.to_numeric(dataframe['LATITUDE'][i])
        long_airport = pd.to_numeric(dataframe['LONGITUDE'][i])
        b = pd.to_numeric(dataframe['Bearing_Degrees'][i])
        d = pd.to_numeric(dataframe['Distance_Kilometers'][i])

        origin = geopy.Point(lat_airport, long_airport)
        destination = geodesic(kilometers=d).destination(origin,b)

        lat2, lon2 = destination.latitude, destination.longitude

        uas_lat.append(lat2)
        uas_long.append(lon2)
        
    
    # Append UAS Lat/Long information to DataFrame
    dataframe['UAS_Latitude'] = uas_lat
    dataframe['UAS_Longitude'] = uas_long
    
    # Export final dataframe (EXCLUDING NULL BEARING INFO) to .csv file
    #dataframe.to_csv('StandardRemark_UAS_LatLong_validBearing_FAAairIdent.csv', index = False)
    
    return dataframe

In [20]:
air227 = uas_lat_long(airports_validDist)
air227

index                                            REMARKS  DATASET  \
0        4  Aircraft observed a 20-30 pound uas pass withi...  AIRPORT   
1        5  Aircraft observed a quad copter uas 500 below ...  AIRPORT   
2        6  Aircraft observed a silver 4 rotor wing uas at...  AIRPORT   
3        7  Aircraft observed a medium sized quadcopter ua...  AIRPORT   
4        8  Aircraft observed a white quad copter type uas...  AIRPORT   
..     ...                                                ...      ...   
222    243  Aircraft reported a white fixed wing uas at 2,...  AIRPORT   
223    244  Aircraft reported a white uas off the left sid...  AIRPORT   
224    246  Aircraft reported a quadcopter uas 500 feet be...  AIRPORT   
225    247  Aircraft reported a uas while SE at 10,000 fee...  AIRPORT   
226    248  Aircraft reported a uas at 8,500 feet 5 W DFW....  AIRPORT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0       1NM SE NEW           1                1.852      SE              135   
1      2 NM SW BJC           2                3.704      SW              225   
2    10 NM NNW VNY          10               18.520     NNW              338   
3        3NM N LIT           3                5.556       N                0   
4     4.5 NM W SLC         4.5                8.334       W              270   
..             ...         ...                  ...     ...              ...   
222   10 NM SW DFW          10               18.520      SW              225   
223  3 NM W of DPA           3                5.556       W              270   
224   7 NM ENE DAL           7               12.964     ENE               68   
225   20 NM SE JFK          20               37.040      SE              135   
226     5 NM W DFW           5                9.260       W              270   

    IDENT                             GLOBAL_ID  \
0     NEW  10BA3990-72A1-4F14-89DB-FC7DF04B54F7   
1     BJC  338AC1EC-7F12-4F3D-932E-E37882849344   
2     VNY  F0AFB4E6-592A-46B2-A76A-5376CC5A0016   
3     LIT  8FBCF8C3-B29D-470D-B4AF-93E8218EDFD5   
4     SLC  CB41ECB2-ADC7-49A5-9C05-81854AFD415C   
..    ...                                   ...   
222   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   
223   DPA  2FD31A12-EF68-4339-891A-D84B47A5EE7B   
224   DAL  48203599-A2E2-467A-BB88-DAC28F3A7FA6   
225   JFK  DD5FB154-0787-4B0F-8554-C033FAFA0155   
226   DFW  B371E693-B91F-40ED-ABB1-628EFE9D3111   

                                       NAME   LATITUDE   LONGITUDE ICAO_ID  \
0                                 Lakefront  30.042417  -90.028250    KNEW   
1                      Rocky Mountain Metro  39.908806 -105.117194    KBJC   
2                                  Van Nuys  34.209806 -118.489972    KVNY   
3    Bill and Hillary Clinton Ntl/Adams Fld  34.729444  -92.224778    KLIT   
4                       Salt Lake City Intl  40.788388 -111.977773    KSLC   
..                                      ...        ...         ...     ...   
222                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW   
223                                  Dupage  41.907053  -88.247995    KDPA   
224                         Dallas Love Fld  32.845945  -96.850877    KDAL   
225                     John F Kennedy Intl  40.639928  -73.778693    KJFK   
226                  Dallas-Fort Worth Intl  32.897233  -97.037695    KDFW   

     UAS_Latitude  UAS_Longitude  
0       30.030602     -90.014673  
1       39.885213    -105.147814  
2       34.364581    -118.565392  
3       34.779527     -92.224778  
4       40.788346    -112.076513  
..            ...            ...  
222     32.779072     -97.177478  
223     41.907034     -88.314958  
224     32.889669     -96.722416  
225     40.403653     -73.470157  
226     32.897194     -97.136667  

[227 rows x 16 columns]

In [21]:
# Export to .csv 
air227.to_csv('airport227_uasLatLong.csv', index = False)

# AIRPORT(?)
`Using the REPORTINGFACILITY field here, so the identifier may be an airport, or may be a navaid or TRACON, etc.`

In [22]:
output471_airportQ = output471_df[output471_df['DATASET'] == 'AIRPORT(?)'].reset_index()
output471_airportQ

index                                            REMARKS     DATASET  \
0      249  Aircraft observed a uas operating 400 feet bel...  AIRPORT(?)   
1      250  Aircraft observed a uas at the same altitude w...  AIRPORT(?)   
2      251  Aircraft observed a uas while SSW bound at 1,4...  AIRPORT(?)   
3      252  Aircraft observed a red uas hovering at 1,000 ...  AIRPORT(?)   
4      253  Aircraft observed a black uas, one foot in dia...  AIRPORT(?)   
..     ...                                                ...         ...   
189    438  Aircraft reported a gray uas while NE bound at...  AIRPORT(?)   
190    439  Aircraft reported a large black fixed wing bla...  AIRPORT(?)   
191    440  Reporting facility unreported. Aircraft report...  AIRPORT(?)   
192    441  Aircraft reported a uas or rocket with a parac...  AIRPORT(?)   
193    442  Aircraft reported a black quadcopter uas while...  AIRPORT(?)   

           UASLOCATION  
0     1.5 NM NW of POC  
1         5NM E of NMM  
2       1 NM NE of SDL  
3         5NM S of D01  
4       5 NM NW of CLE  
..                 ...  
189       5 NM SE  N90  
190        5 NM E  I90  
191   10 NM NE of  SW1  
192      7 NM SSW  A90  
193  4.5 NM SE of  MIC  

[194 rows x 4 columns]

In [23]:
test_airQ = uas_loc_parts(output471_airportQ)
test_airQ

index                                            REMARKS     DATASET  \
0      249  Aircraft observed a uas operating 400 feet bel...  AIRPORT(?)   
1      250  Aircraft observed a uas at the same altitude w...  AIRPORT(?)   
2      251  Aircraft observed a uas while SSW bound at 1,4...  AIRPORT(?)   
3      252  Aircraft observed a red uas hovering at 1,000 ...  AIRPORT(?)   
4      253  Aircraft observed a black uas, one foot in dia...  AIRPORT(?)   
..     ...                                                ...         ...   
189    438  Aircraft reported a gray uas while NE bound at...  AIRPORT(?)   
190    439  Aircraft reported a large black fixed wing bla...  AIRPORT(?)   
191    440  Reporting facility unreported. Aircraft report...  AIRPORT(?)   
192    441  Aircraft reported a uas or rocket with a parac...  AIRPORT(?)   
193    442  Aircraft reported a black quadcopter uas while...  AIRPORT(?)   

           UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0     1.5 NM NW of POC         1.5                2.778      NW   
1         5NM E of NMM           5                9.260       E   
2       1 NM NE of SDL           1                1.852      NE   
3         5NM S of D01           5                9.260       S   
4       5 NM NW of CLE           5                9.260      NW   
..                 ...         ...                  ...     ...   
189       5 NM SE  N90           5                9.260      SE   
190        5 NM E  I90           5                9.260       E   
191   10 NM NE of  SW1          10               18.520      NE   
192      7 NM SSW  A90           7               12.964     SSW   
193  4.5 NM SE of  MIC         4.5                8.334      SE   

     Bearing_Degrees IDENT  
0                315   POC  
1                 90   NMM  
2                 45   SDL  
3                180   D01  
4                315   CLE  
..               ...   ...  
189              135   N90  
190               90   I90  
191               45   SW1  
192              203   A90  
193              135   MIC  

[194 rows x 9 columns]

In [24]:
airq_airlink1 = link_airportsCleaned(test_airQ)
airq_airlink1

index                                            REMARKS     DATASET  \
0      249  Aircraft observed a uas operating 400 feet bel...  AIRPORT(?)   
1      250  Aircraft observed a uas at the same altitude w...  AIRPORT(?)   
2      251  Aircraft observed a uas while SSW bound at 1,4...  AIRPORT(?)   
3      252  Aircraft observed a red uas hovering at 1,000 ...  AIRPORT(?)   
4      253  Aircraft observed a black uas, one foot in dia...  AIRPORT(?)   
..     ...                                                ...         ...   
189    438  Aircraft reported a gray uas while NE bound at...  AIRPORT(?)   
190    439  Aircraft reported a large black fixed wing bla...  AIRPORT(?)   
191    440  Reporting facility unreported. Aircraft report...  AIRPORT(?)   
192    441  Aircraft reported a uas or rocket with a parac...  AIRPORT(?)   
193    442  Aircraft reported a black quadcopter uas while...  AIRPORT(?)   

           UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0     1.5 NM NW of POC         1.5                2.778      NW   
1         5NM E of NMM           5                9.260       E   
2       1 NM NE of SDL           1                1.852      NE   
3         5NM S of D01           5                9.260       S   
4       5 NM NW of CLE           5                9.260      NW   
..                 ...         ...                  ...     ...   
189       5 NM SE  N90           5                9.260      SE   
190        5 NM E  I90           5                9.260       E   
191   10 NM NE of  SW1          10               18.520      NE   
192      7 NM SSW  A90           7               12.964     SSW   
193  4.5 NM SE of  MIC         4.5                8.334      SE   

     Bearing_Degrees IDENT                             GLOBAL_ID  \
0                315   POC  C16DBC19-4892-4450-B04D-F25868853F9E   
1                 90   NMM  D4D40016-A8BC-40D8-A81A-AC03395037FE   
2                 45   SDL  73A4E33E-FACF-472B-860B-37A8BACCDC2C   
3                180   D01                                   NaN   
4                315   CLE  2A962A82-442A-4A82-9535-214B1B978E2E   
..               ...   ...                                   ...   
189              135   N90                                   NaN   
190               90   I90                                   NaN   
191               45   SW1                                   NaN   
192              203   A90                                   NaN   
193              135   MIC  08D02828-E22C-40C0-A0EE-F5A72112C094   

                           NAME   LATITUDE   LONGITUDE ICAO_ID  
0                  Brackett Fld  34.091667 -117.781778    KPOC  
1    Meridian NAS /Mc Cain Fld/  32.553481  -88.555264    KNMM  
2                    Scottsdale  33.622889 -111.910528    KSDL  
3                           NaN        NaN         NaN     NaN  
4        Cleveland-Hopkins Intl  41.409407  -81.854691    KCLE  
..                          ...        ...         ...     ...  
189                         NaN        NaN         NaN     NaN  
190                         NaN        NaN         NaN     NaN  
191                         NaN        NaN         NaN     NaN  
192                         NaN        NaN         NaN     NaN  
193                     Crystal  45.062469  -93.353965    KMIC  

[194 rows x 14 columns]

In [25]:
airq_airlink = airq_airlink1[airq_airlink1['LATITUDE'].notnull()].reset_index(drop=True)
airq_airlink

index                                            REMARKS     DATASET  \
0     249  Aircraft observed a uas operating 400 feet bel...  AIRPORT(?)   
1     250  Aircraft observed a uas at the same altitude w...  AIRPORT(?)   
2     251  Aircraft observed a uas while SSW bound at 1,4...  AIRPORT(?)   
3     253  Aircraft observed a black uas, one foot in dia...  AIRPORT(?)   
4     255  Aircraft observed a blue uas at the same altit...  AIRPORT(?)   
..    ...                                                ...         ...   
94    428  Aircraft observed a uas off the front side whi...  AIRPORT(?)   
95    430  Aircraft observed a uas while E bound at 7,000...  AIRPORT(?)   
96    433  Aircraft reported a red uas 50 feet below thei...  AIRPORT(?)   
97    434  Aircraft reported a uas18 NW of PSPwhileESE bo...  AIRPORT(?)   
98    442  Aircraft reported a black quadcopter uas while...  AIRPORT(?)   

          UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0    1.5 NM NW of POC         1.5                2.778      NW   
1        5NM E of NMM           5                9.260       E   
2      1 NM NE of SDL           1                1.852      NE   
3      5 NM NW of CLE           5                9.260      NW   
4       3 NM S of JAX           3                5.556       S   
..                ...         ...                  ...     ...   
94       6 NM SW  GSB           6               11.112      SW   
95     5 NM W of  PVD           5                9.260       W   
96     8 NM E of  SJC           8               14.816       E   
97   18 NM NW of  PSP          18               33.336      NW   
98  4.5 NM SE of  MIC         4.5                8.334      SE   

    Bearing_Degrees IDENT                             GLOBAL_ID  \
0               315   POC  C16DBC19-4892-4450-B04D-F25868853F9E   
1                90   NMM  D4D40016-A8BC-40D8-A81A-AC03395037FE   
2                45   SDL  73A4E33E-FACF-472B-860B-37A8BACCDC2C   
3               315   CLE  2A962A82-442A-4A82-9535-214B1B978E2E   
4               180   JAX  E2E20781-927D-4930-B08E-27A9C0FE0984   
..              ...   ...                                   ...   
94              225   GSB  B09BE7CE-8921-4760-8DF3-B7226DD0EE65   
95              270   PVD  3F706D32-8D17-40DB-8A06-71C13F2158F1   
96               90   SJC  ACA1CDD6-3606-47F8-9A7B-904DB1B3DBC5   
97              315   PSP  D207DB5C-CC14-49EC-BFC7-9D5BFC1F9979   
98              135   MIC  08D02828-E22C-40C0-A0EE-F5A72112C094   

                             NAME   LATITUDE   LONGITUDE ICAO_ID  
0                    Brackett Fld  34.091667 -117.781778    KPOC  
1      Meridian NAS /Mc Cain Fld/  32.553481  -88.555264    KNMM  
2                      Scottsdale  33.622889 -111.910528    KSDL  
3          Cleveland-Hopkins Intl  41.409407  -81.854691    KCLE  
4               Jacksonville Intl  30.494046  -81.687847    KJAX  
..                            ...        ...         ...     ...  
94            Seymour Johnson AFB  35.339385  -77.960594    KGSB  
95   Theodore Francis Green State  41.722333  -71.427722    KPVD  
96  Norman Y Mineta San Jose Intl  37.362995 -121.928621    KSJC  
97              Palm Springs Intl  33.829667 -116.506694    KPSP  
98                        Crystal  45.062469  -93.353965    KMIC  

[99 rows x 14 columns]

In [26]:
air99 = uas_lat_long(airq_airlink)
air99

index                                            REMARKS     DATASET  \
0     249  Aircraft observed a uas operating 400 feet bel...  AIRPORT(?)   
1     250  Aircraft observed a uas at the same altitude w...  AIRPORT(?)   
2     251  Aircraft observed a uas while SSW bound at 1,4...  AIRPORT(?)   
3     253  Aircraft observed a black uas, one foot in dia...  AIRPORT(?)   
4     255  Aircraft observed a blue uas at the same altit...  AIRPORT(?)   
..    ...                                                ...         ...   
94    428  Aircraft observed a uas off the front side whi...  AIRPORT(?)   
95    430  Aircraft observed a uas while E bound at 7,000...  AIRPORT(?)   
96    433  Aircraft reported a red uas 50 feet below thei...  AIRPORT(?)   
97    434  Aircraft reported a uas18 NW of PSPwhileESE bo...  AIRPORT(?)   
98    442  Aircraft reported a black quadcopter uas while...  AIRPORT(?)   

          UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0    1.5 NM NW of POC         1.5                2.778      NW   
1        5NM E of NMM           5                9.260       E   
2      1 NM NE of SDL           1                1.852      NE   
3      5 NM NW of CLE           5                9.260      NW   
4       3 NM S of JAX           3                5.556       S   
..                ...         ...                  ...     ...   
94       6 NM SW  GSB           6               11.112      SW   
95     5 NM W of  PVD           5                9.260       W   
96     8 NM E of  SJC           8               14.816       E   
97   18 NM NW of  PSP          18               33.336      NW   
98  4.5 NM SE of  MIC         4.5                8.334      SE   

    Bearing_Degrees IDENT                             GLOBAL_ID  \
0               315   POC  C16DBC19-4892-4450-B04D-F25868853F9E   
1                90   NMM  D4D40016-A8BC-40D8-A81A-AC03395037FE   
2                45   SDL  73A4E33E-FACF-472B-860B-37A8BACCDC2C   
3               315   CLE  2A962A82-442A-4A82-9535-214B1B978E2E   
4               180   JAX  E2E20781-927D-4930-B08E-27A9C0FE0984   
..              ...   ...                                   ...   
94              225   GSB  B09BE7CE-8921-4760-8DF3-B7226DD0EE65   
95              270   PVD  3F706D32-8D17-40DB-8A06-71C13F2158F1   
96               90   SJC  ACA1CDD6-3606-47F8-9A7B-904DB1B3DBC5   
97              315   PSP  D207DB5C-CC14-49EC-BFC7-9D5BFC1F9979   
98              135   MIC  08D02828-E22C-40C0-A0EE-F5A72112C094   

                             NAME   LATITUDE   LONGITUDE ICAO_ID  \
0                    Brackett Fld  34.091667 -117.781778    KPOC   
1      Meridian NAS /Mc Cain Fld/  32.553481  -88.555264    KNMM   
2                      Scottsdale  33.622889 -111.910528    KSDL   
3          Cleveland-Hopkins Intl  41.409407  -81.854691    KCLE   
4               Jacksonville Intl  30.494046  -81.687847    KJAX   
..                            ...        ...         ...     ...   
94            Seymour Johnson AFB  35.339385  -77.960594    KGSB   
95   Theodore Francis Green State  41.722333  -71.427722    KPVD   
96  Norman Y Mineta San Jose Intl  37.362995 -121.928621    KSJC   
97              Palm Springs Intl  33.829667 -116.506694    KPSP   
98                        Crystal  45.062469  -93.353965    KMIC   

    UAS_Latitude  UAS_Longitude  
0      34.109374    -117.803068  
1      32.553442     -88.456671  
2      33.634695    -111.896413  
3      41.468336     -81.933073  
4      30.443928     -81.687847  
..           ...            ...  
94     35.268532     -78.046950  
95     41.722279     -71.539007  
96     37.362876    -121.761372  
97     34.041915    -116.761971  
98     45.009418     -93.279213  

[99 rows x 16 columns]

In [27]:
# Export to .csv 
air99.to_csv('air99_uasLatLong.csv', index = False)

In [28]:
airq_airNOlink = airq_airlink1[airq_airlink1['LATITUDE'].isnull()].reset_index(drop=True)
airq_airNOlink

index                                            REMARKS     DATASET  \
0     252  Aircraft observed a red uas hovering at 1,000 ...  AIRPORT(?)   
1     254  Aircraft observed a uas off their left side wh...  AIRPORT(?)   
2     256  Aircraft observed a uas 700 feet below while w...  AIRPORT(?)   
3     257  Aircraft observed a uas E bound at 9,300 feet ...  AIRPORT(?)   
4     258  Aircraft observed a silver uas while NE bound ...  AIRPORT(?)   
..    ...                                                ...         ...   
90    437  Aircraft reported a silver quad copter at 6,50...  AIRPORT(?)   
91    438  Aircraft reported a gray uas while NE bound at...  AIRPORT(?)   
92    439  Aircraft reported a large black fixed wing bla...  AIRPORT(?)   
93    440  Reporting facility unreported. Aircraft report...  AIRPORT(?)   
94    441  Aircraft reported a uas or rocket with a parac...  AIRPORT(?)   

         UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0       5NM S of D01           5                9.260       S   
1    10NM ENE of F11          10               18.520     ENE   
2     5 NM NW of ZME           5                9.260      NW   
3     5 NM NW of M03           5                9.260      NW   
4       1NM S of WOC           1                1.852       S   
..               ...         ...                  ...     ...   
90   9 NM NW of  C90           9               16.668      NW   
91      5 NM SE  N90           5                9.260      SE   
92       5 NM E  I90           5                9.260       E   
93  10 NM NE of  SW1          10               18.520      NE   
94     7 NM SSW  A90           7               12.964     SSW   

    Bearing_Degrees IDENT GLOBAL_ID NAME  LATITUDE  LONGITUDE ICAO_ID  
0               180   D01       NaN  NaN       NaN        NaN     NaN  
1                68   F11       NaN  NaN       NaN        NaN     NaN  
2               315   ZME       NaN  NaN       NaN        NaN     NaN  
3               315   M03       NaN  NaN       NaN        NaN     NaN  
4               180   WOC       NaN  NaN       NaN        NaN     NaN  
..              ...   ...       ...  ...       ...        ...     ...  
90              315   C90       NaN  NaN       NaN        NaN     NaN  
91              135   N90       NaN  NaN       NaN        NaN     NaN  
92               90   I90       NaN  NaN       NaN        NaN     NaN  
93               45   SW1       NaN  NaN       NaN        NaN     NaN  
94              203   A90       NaN  NaN       NaN        NaN     NaN  

[95 rows x 14 columns]

In [53]:
noair_nonav = pd.merge(airq_airNOlink, output471, on = 'REMARKS', how = 'left')
noair_nonav

index_x                                            REMARKS   DATASET_x  \
0       252  Aircraft observed a red uas hovering at 1,000 ...  AIRPORT(?)   
1       254  Aircraft observed a uas off their left side wh...  AIRPORT(?)   
2       256  Aircraft observed a uas 700 feet below while w...  AIRPORT(?)   
3       257  Aircraft observed a uas E bound at 9,300 feet ...  AIRPORT(?)   
4       258  Aircraft observed a silver uas while NE bound ...  AIRPORT(?)   
..      ...                                                ...         ...   
90      437  Aircraft reported a silver quad copter at 6,50...  AIRPORT(?)   
91      438  Aircraft reported a gray uas while NE bound at...  AIRPORT(?)   
92      439  Aircraft reported a large black fixed wing bla...  AIRPORT(?)   
93      440  Reporting facility unreported. Aircraft report...  AIRPORT(?)   
94      441  Aircraft reported a uas or rocket with a parac...  AIRPORT(?)   

       UASLOCATION_x Distance_NM  Distance_Kilometers Bearing  \
0       5NM S of D01           5                9.260       S   
1    10NM ENE of F11          10               18.520     ENE   
2     5 NM NW of ZME           5                9.260      NW   
3     5 NM NW of M03           5                9.260      NW   
4       1NM S of WOC           1                1.852       S   
..               ...         ...                  ...     ...   
90   9 NM NW of  C90           9               16.668      NW   
91      5 NM SE  N90           5                9.260      SE   
92       5 NM E  I90           5                9.260       E   
93  10 NM NE of  SW1          10               18.520      NE   
94     7 NM SSW  A90           7               12.964     SSW   

    Bearing_Degrees IDENT GLOBAL_ID  ... UASWEIGHTGT55.CEDAR  UASDIM.CEDAR  \
0               180   D01       NaN  ...                 NaN           NaN   
1                68   F11       NaN  ...                 NaN           NaN   
2               315   ZME       NaN  ...                 NaN           NaN   
3               315   M03       NaN  ...                 NaN           NaN   
4               180   WOC       NaN  ...                 NaN           NaN   
..              ...   ...       ...  ...                 ...           ...   
90              315   C90       NaN  ...                 NaN           NaN   
91              135   N90       NaN  ...                 NaN           NaN   
92               90   I90       NaN  ...                 NaN           NaN   
93               45   SW1       NaN  ...                 NaN           NaN   
94              203   A90       NaN  ...                 NaN           NaN   

    UASFWROTOR.CEDAR PILOTREPORTEDNMAC.CEDAR  TCASRA.CEDAR SUMMARY.CEDAR  \
0                NaN                     NaN           NaN           NaN   
1                NaN                     NaN           NaN           NaN   
2                NaN                     NaN           NaN           NaN   
3                NaN                     NaN           NaN           NaN   
4                NaN                     NaN           NaN           NaN   
..               ...                     ...           ...           ...   
90               NaN                     NaN           NaN           NaN   
91               NaN                     NaN           NaN           NaN   
92               NaN                     NaN           NaN           NaN   
93               NaN                     NaN           NaN           NaN   
94               NaN                     NaN           NaN           NaN   

   QAFINDINGS.CEDAR   DATASET_y     UASLOCATION_y RWYLOCATION  
0               NaN  AIRPORT(?)      5NM S of D01         NaN  
1               NaN  AIRPORT(?)   10NM ENE of F11         NaN  
2               NaN  AIRPORT(?)    5 NM NW of ZME         NaN  
3               NaN  AIRPORT(?)    5 NM NW of M03         NaN  
4               NaN  AIRPORT(?)      1NM S of WOC         NaN  
..              ...         ...               ...   

In [55]:
# Export to .csv 
noair_nonav.to_csv('noAirNav95OF471_EXCEPTIONS.csv', index = False)

In [29]:
#test_airnav = link_navaidCleaned(airq_airNOlink)
#test_airnav

# NAVAID Systems

In [30]:
output471_nav = output471_df[output471_df['DATASET'] == 'NAVAID'].reset_index()
output471_nav

index                                            REMARKS DATASET  \
0      0  Aircraft observed a large uas 2-5 NM away at h...  NAVAID   
1      1  Aircraft reported an orange quad copter uas 3 ...  NAVAID   
2      2  Aircraft reported an orange quad copter uas mo...  NAVAID   

    UASLOCATION  
0   21NM SW ANX  
1   26 NM S LTJ  
2  18NM ESE TNV

In [31]:
test_nav = uas_loc_parts(output471_nav)
test_nav

index                                            REMARKS DATASET  \
0      0  Aircraft observed a large uas 2-5 NM away at h...  NAVAID   
1      1  Aircraft reported an orange quad copter uas 3 ...  NAVAID   
2      2  Aircraft reported an orange quad copter uas mo...  NAVAID   

    UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0   21NM SW ANX          21               38.892      SW              225   
1   26 NM S LTJ          26               48.152       S              180   
2  18NM ESE TNV          18               33.336     ESE              113   

  IDENT  
0   ANX  
1   LTJ  
2   TNV

In [32]:
def link_navaidCleaned(dataframe):
    '''
    This function links the provided dataframe to the NAVAID_cleaned.csv file.
    
    input : dataframe is a df with an 'IDENT' field on which we will link to the airports_cleaned.csv
    output : the dataframe provided as input with navaid lat/long fields appended
    '''
    
    # Read in Cleaned Airports dataset
    navC = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/NAVAID_cleaned.csv')
    
    return pd.merge(dataframe, navC, on='IDENT', how='left')

In [33]:
link_navaidCleaned(test_nav)

index                                            REMARKS DATASET  \
0      0  Aircraft observed a large uas 2-5 NM away at h...  NAVAID   
1      1  Aircraft reported an orange quad copter uas 3 ...  NAVAID   
2      2  Aircraft reported an orange quad copter uas mo...  NAVAID   

    UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0   21NM SW ANX          21               38.892      SW              225   
1   26 NM S LTJ          26               48.152       S              180   
2  18NM ESE TNV          18               33.336     ESE              113   

  IDENT   LONGITUDE   LATITUDE   NAME_TXT        CITY  
0   ANX  -94.128842  39.095431   NAPOLEON    NAPOLEON  
1   LTJ -121.100848  45.713589  KLICKITAT  THE DALLES  
2   TNV  -96.058247  30.288532   NAVASOTA    NAVASOTA

In [34]:
nav3_latLong = uas_lat_long(link_navaidCleaned(test_nav))
nav3_latLong

index                                            REMARKS DATASET  \
0      0  Aircraft observed a large uas 2-5 NM away at h...  NAVAID   
1      1  Aircraft reported an orange quad copter uas 3 ...  NAVAID   
2      2  Aircraft reported an orange quad copter uas mo...  NAVAID   

    UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0   21NM SW ANX          21               38.892      SW              225   
1   26 NM S LTJ          26               48.152       S              180   
2  18NM ESE TNV          18               33.336     ESE              113   

  IDENT   LONGITUDE   LATITUDE   NAME_TXT        CITY  UAS_Latitude  \
0   ANX  -94.128842  39.095431   NAPOLEON    NAPOLEON     38.847278   
1   LTJ -121.100848  45.713589  KLICKITAT  THE DALLES     45.280339   
2   TNV  -96.058247  30.288532   NAVASOTA    NAVASOTA     30.170646   

   UAS_Longitude  
0     -94.445626  
1    -121.100848  
2     -95.739666

In [35]:
# Export to .csv 
nav3_latLong.to_csv('nav3_uasLatLong.csv', index = False)

# DESIGNATED POINTS

In [36]:
output471_dp = output471_df[output471_df['DATASET'] == 'DESIGNATED POINT'].reset_index()
output471_dp

index                                            REMARKS           DATASET  \
0      3  Aircraft observed a uas while S bound at 2,000...  DESIGNATED POINT   

       UASLOCATION  
0  2 NM N of TUGGZ

In [37]:
uas_loc_parts(output471_dp)

index                                            REMARKS           DATASET  \
0      3  Aircraft observed a uas while S bound at 2,000...  DESIGNATED POINT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0  2 NM N of TUGGZ           2                3.704       N                0   

   IDENT  
0  TUGGZ

In [38]:
def dp_link (dataframe):
    '''
    This functions merges the dataframe input by the user with the cleaned NAVAID_cleaned.csv dataset
    '''
    
    # Read in Cleaned NAVAID Systems dataset
    dp_clean = pd.read_csv('C:/Users/grace/OneDrive/Desktop/GMU/DAEN690/designated_points_cleaned.csv')
    
    output_df = pd.merge(dataframe, dp_clean, on = 'IDENT', how = 'left').reset_index(drop=True)
    
    return output_df

In [39]:
dp_link(uas_loc_parts(output471_dp))

index                                            REMARKS           DATASET  \
0      3  Aircraft observed a uas while S bound at 2,000...  DESIGNATED POINT   

       UASLOCATION Distance_NM  Distance_Kilometers Bearing  Bearing_Degrees  \
0  2 NM N of TUGGZ           2                3.704       N                0   

   IDENT  Unnamed: 0  LATITUDE  LONGITUDE  
0  TUGGZ         NaN       NaN        NaN

In [40]:
# Export to .csv 
output471_dp.to_csv('dp1OF471_EXCEPTIONS.csv', index = False)

# FIXED RADIAL DISTANCES

In [41]:
output471_fr = output471_df[output471_df['DATASET'] == 'FIXED RADIAL'].reset_index()
output471_fr

index                                            REMARKS       DATASET  \
0     443  Aircraft observed a uas south bound at 2,500 f...  FIXED RADIAL   
1     444  DCA-1: Aircraft observed a uas NW of the DC Ch...  FIXED RADIAL   
2     445  NEMAC: Aircraft observed a quad-copter uas wit...  FIXED RADIAL   
3     446  Aircraft observed a Red and Black uas at the A...  FIXED RADIAL   
4     447  Aircraft observed a uas at approximately the B...  FIXED RADIAL   
5     448  Aircraft observed a uas off the DCA205017 alti...  FIXED RADIAL   
6     449  Aircraft observed a uas off the front while NW...  FIXED RADIAL   
7     450  Aircraft observed a uas 1,500 feet below them ...  FIXED RADIAL   
8     451  Aircraft observed a blue uas overhead while W ...  FIXED RADIAL   
9     452  Aircraft observed a uas off the left side whil...  FIXED RADIAL   
10    453  Aircraft observed a uas at 600 feet off the le...  FIXED RADIAL   
11    454  Aircraft observed a uas off the right side whi...  FIXED RADIAL   
12    455  DCA-1: Aircraft observed a uas off the front s...  FIXED RADIAL   
13    456  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
14    457  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
15    458  Aircraft observed a uas E bound at 16,000 feet...  FIXED RADIAL   
16    459  Aircraft observed a uas while E bound at FL240...  FIXED RADIAL   
17    460  Aircraft observed a uas (quadcopter) off the s...  FIXED RADIAL   
18    461  Aircraft observed a yellow and red uas off the...  FIXED RADIAL   
19    462  Aircraft observed a red uas, approximately 3-4...  FIXED RADIAL   
20    463  Aircraft observed a black uas off the left sid...  FIXED RADIAL   
21    464  DCA-1 Aircraft observed a red/silver uas off t...  FIXED RADIAL   
22    465  Aircraft observed a white quadcopter uas off t...  FIXED RADIAL   
23    466  Aircraft observed a uas  while northbound at 2...  FIXED RADIAL   
24    467  Aircraft reported a yellow drone approximately...  FIXED RADIAL   
25    468  Aircraft reported a small yellow drone DCA0660...  FIXED RADIAL   
26    469  DCA-2: Aircraft reported a blue and red uas at...  FIXED RADIAL   
27    470  Aircraft reported a white quad copter at the B...  FIXED RADIAL   

          UASLOCATION  
0   009 NM 295 of RIC  
1   004 NM 024 of DCA  
2   010 NM 300 of VCV  
3   008 NM 140 of AML  
4   025 NM 171 of BWI  
5   017 NM 205 of DCA  
6   037 NM 107 of SXC  
7   015 NM 142 of SAX  
8   009 NM 226 of CLT  
9   006 NM 080 of RDU  
10  007 NM 023 of BWI  
11  038 NM 297 of DCA  
12  010 NM 038 of DCA  
13  064 NM 297 of PHX  
14  010 NM 339 of PHX  
15  046 NM 248 of GBN  
16  051 NM 022 of GBN  
17  008 NM 104 of BWI  
18  008 NM 241 of IAD  
19  023 NM 160 of BWI  
20  025 NM 160 of DCA  
21  012 NM 281 of DCA  
22  005 NM 120 of BAL  
23  012 NM 030 of AGC  
24  012 NM 036 of DCA  
25  030 NM 066 of DCA  
26  008 NM 201 of DCA  
27  010 NM 320 of BRV

In [42]:
test_fr = uas_loc_parts(output471_fr)
test_fr

index                                            REMARKS       DATASET  \
0     443  Aircraft observed a uas south bound at 2,500 f...  FIXED RADIAL   
1     444  DCA-1: Aircraft observed a uas NW of the DC Ch...  FIXED RADIAL   
2     445  NEMAC: Aircraft observed a quad-copter uas wit...  FIXED RADIAL   
3     446  Aircraft observed a Red and Black uas at the A...  FIXED RADIAL   
4     447  Aircraft observed a uas at approximately the B...  FIXED RADIAL   
5     448  Aircraft observed a uas off the DCA205017 alti...  FIXED RADIAL   
6     449  Aircraft observed a uas off the front while NW...  FIXED RADIAL   
7     450  Aircraft observed a uas 1,500 feet below them ...  FIXED RADIAL   
8     451  Aircraft observed a blue uas overhead while W ...  FIXED RADIAL   
9     452  Aircraft observed a uas off the left side whil...  FIXED RADIAL   
10    453  Aircraft observed a uas at 600 feet off the le...  FIXED RADIAL   
11    454  Aircraft observed a uas off the right side whi...  FIXED RADIAL   
12    455  DCA-1: Aircraft observed a uas off the front s...  FIXED RADIAL   
13    456  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
14    457  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
15    458  Aircraft observed a uas E bound at 16,000 feet...  FIXED RADIAL   
16    459  Aircraft observed a uas while E bound at FL240...  FIXED RADIAL   
17    460  Aircraft observed a uas (quadcopter) off the s...  FIXED RADIAL   
18    461  Aircraft observed a yellow and red uas off the...  FIXED RADIAL   
19    462  Aircraft observed a red uas, approximately 3-4...  FIXED RADIAL   
20    463  Aircraft observed a black uas off the left sid...  FIXED RADIAL   
21    464  DCA-1 Aircraft observed a red/silver uas off t...  FIXED RADIAL   
22    465  Aircraft observed a white quadcopter uas off t...  FIXED RADIAL   
23    466  Aircraft observed a uas  while northbound at 2...  FIXED RADIAL   
24    467  Aircraft reported a yellow drone approximately...  FIXED RADIAL   
25    468  Aircraft reported a small yellow drone DCA0660...  FIXED RADIAL   
26    469  DCA-2: Aircraft reported a blue and red uas at...  FIXED RADIAL   
27    470  Aircraft reported a white quad copter at the B...  FIXED RADIAL   

          UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0   009 NM 295 of RIC         009               16.668     295   
1   004 NM 024 of DCA         004                7.408     024   
2   010 NM 300 of VCV         010               18.520     300   
3   008 NM 140 of AML         008               14.816     140   
4   025 NM 171 of BWI         025               46.300     171   
5   017 NM 205 of DCA         017               31.484     205   
6   037 NM 107 of SXC         037               68.524     107   
7   015 NM 142 of SAX         015               27.780     142   
8   009 NM 226 of CLT         009               16.668     226   
9   006 NM 080 of RDU         006               11.112     080   
10  007 NM 023 of BWI         007               12.964     023   
11  038 NM 297 of DCA         038               70.376     297   
12  010 NM 038 of DCA         010               18.520     038   
13  064 NM 297 of PHX         064              118.528     297   
14  010 NM 339 of PHX         010               18.520     339   
15  046 NM 248 of GBN         046               85.192     248   
16  051 NM 022 of GBN         051               94.452     022   
17  008 NM 104 of BWI         008               14.816     104   
18  008 NM 241 of IAD         008               14.816     241   
19  023 NM 160 of BWI         023               42.596     160   
20  025 NM 160 of DCA         025               46.300     160   
21  012 NM 281 of DCA         012               22.224     281   
22  005 NM 120 of BAL         005                9.260     120   
23  012 NM 030 of AGC         012               22.224     030   
24  012 NM 036 of DCA         012               22.224     036   
25  030 NM 066 of DCA    

In [43]:
fr_airlink1 = link_airportsCleaned(test_fr)
fr_airlink1

index                                            REMARKS       DATASET  \
0     443  Aircraft observed a uas south bound at 2,500 f...  FIXED RADIAL   
1     444  DCA-1: Aircraft observed a uas NW of the DC Ch...  FIXED RADIAL   
2     445  NEMAC: Aircraft observed a quad-copter uas wit...  FIXED RADIAL   
3     446  Aircraft observed a Red and Black uas at the A...  FIXED RADIAL   
4     447  Aircraft observed a uas at approximately the B...  FIXED RADIAL   
5     448  Aircraft observed a uas off the DCA205017 alti...  FIXED RADIAL   
6     449  Aircraft observed a uas off the front while NW...  FIXED RADIAL   
7     450  Aircraft observed a uas 1,500 feet below them ...  FIXED RADIAL   
8     451  Aircraft observed a blue uas overhead while W ...  FIXED RADIAL   
9     452  Aircraft observed a uas off the left side whil...  FIXED RADIAL   
10    453  Aircraft observed a uas at 600 feet off the le...  FIXED RADIAL   
11    454  Aircraft observed a uas off the right side whi...  FIXED RADIAL   
12    455  DCA-1: Aircraft observed a uas off the front s...  FIXED RADIAL   
13    456  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
14    457  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
15    458  Aircraft observed a uas E bound at 16,000 feet...  FIXED RADIAL   
16    459  Aircraft observed a uas while E bound at FL240...  FIXED RADIAL   
17    460  Aircraft observed a uas (quadcopter) off the s...  FIXED RADIAL   
18    461  Aircraft observed a yellow and red uas off the...  FIXED RADIAL   
19    462  Aircraft observed a red uas, approximately 3-4...  FIXED RADIAL   
20    463  Aircraft observed a black uas off the left sid...  FIXED RADIAL   
21    464  DCA-1 Aircraft observed a red/silver uas off t...  FIXED RADIAL   
22    465  Aircraft observed a white quadcopter uas off t...  FIXED RADIAL   
23    466  Aircraft observed a uas  while northbound at 2...  FIXED RADIAL   
24    467  Aircraft reported a yellow drone approximately...  FIXED RADIAL   
25    468  Aircraft reported a small yellow drone DCA0660...  FIXED RADIAL   
26    469  DCA-2: Aircraft reported a blue and red uas at...  FIXED RADIAL   
27    470  Aircraft reported a white quad copter at the B...  FIXED RADIAL   

          UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0   009 NM 295 of RIC         009               16.668     295   
1   004 NM 024 of DCA         004                7.408     024   
2   010 NM 300 of VCV         010               18.520     300   
3   008 NM 140 of AML         008               14.816     140   
4   025 NM 171 of BWI         025               46.300     171   
5   017 NM 205 of DCA         017               31.484     205   
6   037 NM 107 of SXC         037               68.524     107   
7   015 NM 142 of SAX         015               27.780     142   
8   009 NM 226 of CLT         009               16.668     226   
9   006 NM 080 of RDU         006               11.112     080   
10  007 NM 023 of BWI         007               12.964     023   
11  038 NM 297 of DCA         038               70.376     297   
12  010 NM 038 of DCA         010               18.520     038   
13  064 NM 297 of PHX         064              118.528     297   
14  010 NM 339 of PHX         010               18.520     339   
15  046 NM 248 of GBN         046               85.192     248   
16  051 NM 022 of GBN         051               94.452     022   
17  008 NM 104 of BWI         008               14.816     104   
18  008 NM 241 of IAD         008               14.816     241   
19  023 NM 160 of BWI         023               42.596     160   
20  025 NM 160 of DCA         025               46.300     160   
21  012 NM 281 of DCA         012               22.224     281   
22  005 NM 120 of BAL         005                9.260     120   
23  012 NM 030 of AGC         012               22.224     030   
24  012 NM 036 of DCA         012               22.224     036   
25  030 NM 066 of DCA    

In [44]:
fr_airlink = fr_airlink1[fr_airlink1['LATITUDE'].notnull()].reset_index(drop=True)
fr_airlink

index                                            REMARKS       DATASET  \
0     443  Aircraft observed a uas south bound at 2,500 f...  FIXED RADIAL   
1     444  DCA-1: Aircraft observed a uas NW of the DC Ch...  FIXED RADIAL   
2     445  NEMAC: Aircraft observed a quad-copter uas wit...  FIXED RADIAL   
3     447  Aircraft observed a uas at approximately the B...  FIXED RADIAL   
4     448  Aircraft observed a uas off the DCA205017 alti...  FIXED RADIAL   
5     451  Aircraft observed a blue uas overhead while W ...  FIXED RADIAL   
6     452  Aircraft observed a uas off the left side whil...  FIXED RADIAL   
7     453  Aircraft observed a uas at 600 feet off the le...  FIXED RADIAL   
8     454  Aircraft observed a uas off the right side whi...  FIXED RADIAL   
9     455  DCA-1: Aircraft observed a uas off the front s...  FIXED RADIAL   
10    456  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
11    457  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
12    460  Aircraft observed a uas (quadcopter) off the s...  FIXED RADIAL   
13    461  Aircraft observed a yellow and red uas off the...  FIXED RADIAL   
14    462  Aircraft observed a red uas, approximately 3-4...  FIXED RADIAL   
15    463  Aircraft observed a black uas off the left sid...  FIXED RADIAL   
16    464  DCA-1 Aircraft observed a red/silver uas off t...  FIXED RADIAL   
17    466  Aircraft observed a uas  while northbound at 2...  FIXED RADIAL   
18    467  Aircraft reported a yellow drone approximately...  FIXED RADIAL   
19    468  Aircraft reported a small yellow drone DCA0660...  FIXED RADIAL   
20    469  DCA-2: Aircraft reported a blue and red uas at...  FIXED RADIAL   

          UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0   009 NM 295 of RIC         009               16.668     295   
1   004 NM 024 of DCA         004                7.408     024   
2   010 NM 300 of VCV         010               18.520     300   
3   025 NM 171 of BWI         025               46.300     171   
4   017 NM 205 of DCA         017               31.484     205   
5   009 NM 226 of CLT         009               16.668     226   
6   006 NM 080 of RDU         006               11.112     080   
7   007 NM 023 of BWI         007               12.964     023   
8   038 NM 297 of DCA         038               70.376     297   
9   010 NM 038 of DCA         010               18.520     038   
10  064 NM 297 of PHX         064              118.528     297   
11  010 NM 339 of PHX         010               18.520     339   
12  008 NM 104 of BWI         008               14.816     104   
13  008 NM 241 of IAD         008               14.816     241   
14  023 NM 160 of BWI         023               42.596     160   
15  025 NM 160 of DCA         025               46.300     160   
16  012 NM 281 of DCA         012               22.224     281   
17  012 NM 030 of AGC         012               22.224     030   
18  012 NM 036 of DCA         012               22.224     036   
19  030 NM 066 of DCA         030               55.560     066   
20  008 NM 201 of DCA         008               14.816     201   

   Bearing_Degrees IDENT                             GLOBAL_ID  \
0              295   RIC  69E61D8F-915F-4B1D-9666-6338F1F19CD9   
1              024   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
2              300   VCV  D0BF5C97-F023-4A00-A9DE-B7317305C810   
3              171   BWI  751671DD-2768-454F-AA80-2567DB9933FE   
4              205   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
5              226   CLT  91CC8955-7A21-4843-9C70-CC0903339770   
6              080   RDU  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4   
7              023   BWI  751671DD-2768-454F-AA80-2567DB9933FE   
8              297   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
9              038   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
10             297   PHX  639E89B2-F2AE-4589-BC8E-2694490CE138   
11             339   PHX  639E89B2-F2AE-45

In [45]:
fix_radial21 = uas_lat_long(fr_airlink)
fix_radial21

index                                            REMARKS       DATASET  \
0     443  Aircraft observed a uas south bound at 2,500 f...  FIXED RADIAL   
1     444  DCA-1: Aircraft observed a uas NW of the DC Ch...  FIXED RADIAL   
2     445  NEMAC: Aircraft observed a quad-copter uas wit...  FIXED RADIAL   
3     447  Aircraft observed a uas at approximately the B...  FIXED RADIAL   
4     448  Aircraft observed a uas off the DCA205017 alti...  FIXED RADIAL   
5     451  Aircraft observed a blue uas overhead while W ...  FIXED RADIAL   
6     452  Aircraft observed a uas off the left side whil...  FIXED RADIAL   
7     453  Aircraft observed a uas at 600 feet off the le...  FIXED RADIAL   
8     454  Aircraft observed a uas off the right side whi...  FIXED RADIAL   
9     455  DCA-1: Aircraft observed a uas off the front s...  FIXED RADIAL   
10    456  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
11    457  Aircraft observed a white quad uas while E bou...  FIXED RADIAL   
12    460  Aircraft observed a uas (quadcopter) off the s...  FIXED RADIAL   
13    461  Aircraft observed a yellow and red uas off the...  FIXED RADIAL   
14    462  Aircraft observed a red uas, approximately 3-4...  FIXED RADIAL   
15    463  Aircraft observed a black uas off the left sid...  FIXED RADIAL   
16    464  DCA-1 Aircraft observed a red/silver uas off t...  FIXED RADIAL   
17    466  Aircraft observed a uas  while northbound at 2...  FIXED RADIAL   
18    467  Aircraft reported a yellow drone approximately...  FIXED RADIAL   
19    468  Aircraft reported a small yellow drone DCA0660...  FIXED RADIAL   
20    469  DCA-2: Aircraft reported a blue and red uas at...  FIXED RADIAL   

          UASLOCATION Distance_NM  Distance_Kilometers Bearing  \
0   009 NM 295 of RIC         009               16.668     295   
1   004 NM 024 of DCA         004                7.408     024   
2   010 NM 300 of VCV         010               18.520     300   
3   025 NM 171 of BWI         025               46.300     171   
4   017 NM 205 of DCA         017               31.484     205   
5   009 NM 226 of CLT         009               16.668     226   
6   006 NM 080 of RDU         006               11.112     080   
7   007 NM 023 of BWI         007               12.964     023   
8   038 NM 297 of DCA         038               70.376     297   
9   010 NM 038 of DCA         010               18.520     038   
10  064 NM 297 of PHX         064              118.528     297   
11  010 NM 339 of PHX         010               18.520     339   
12  008 NM 104 of BWI         008               14.816     104   
13  008 NM 241 of IAD         008               14.816     241   
14  023 NM 160 of BWI         023               42.596     160   
15  025 NM 160 of DCA         025               46.300     160   
16  012 NM 281 of DCA         012               22.224     281   
17  012 NM 030 of AGC         012               22.224     030   
18  012 NM 036 of DCA         012               22.224     036   
19  030 NM 066 of DCA         030               55.560     066   
20  008 NM 201 of DCA         008               14.816     201   

   Bearing_Degrees IDENT                             GLOBAL_ID  \
0              295   RIC  69E61D8F-915F-4B1D-9666-6338F1F19CD9   
1              024   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
2              300   VCV  D0BF5C97-F023-4A00-A9DE-B7317305C810   
3              171   BWI  751671DD-2768-454F-AA80-2567DB9933FE   
4              205   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
5              226   CLT  91CC8955-7A21-4843-9C70-CC0903339770   
6              080   RDU  CB0BC5B9-0B40-44DE-96E3-5B8E2AEA35C4   
7              023   BWI  751671DD-2768-454F-AA80-2567DB9933FE   
8              297   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
9              038   DCA  96204F8C-A545-4DF9-943D-5472C5BC1697   
10             297   PHX  639E89B2-F2AE-4589-BC8E-2694490CE138   
11             339   PHX  639E89B2-F2AE-45

In [46]:
# Export to .csv 
fix_radial21.to_csv('fixRadial21_uasLatLong.csv', index = False)

In [47]:
fr_airNONlink = fr_airlink1[fr_airlink1['LATITUDE'].isnull()].reset_index(drop=True).drop(columns = ['GLOBAL_ID', 'NAME', 'LATITUDE', 'LONGITUDE', 'ICAO_ID'])
fr_airNONlink

index                                            REMARKS       DATASET  \
0    446  Aircraft observed a Red and Black uas at the A...  FIXED RADIAL   
1    449  Aircraft observed a uas off the front while NW...  FIXED RADIAL   
2    450  Aircraft observed a uas 1,500 feet below them ...  FIXED RADIAL   
3    458  Aircraft observed a uas E bound at 16,000 feet...  FIXED RADIAL   
4    459  Aircraft observed a uas while E bound at FL240...  FIXED RADIAL   
5    465  Aircraft observed a white quadcopter uas off t...  FIXED RADIAL   
6    470  Aircraft reported a white quad copter at the B...  FIXED RADIAL   

         UASLOCATION Distance_NM  Distance_Kilometers Bearing Bearing_Degrees  \
0  008 NM 140 of AML         008               14.816     140             140   
1  037 NM 107 of SXC         037               68.524     107             107   
2  015 NM 142 of SAX         015               27.780     142             142   
3  046 NM 248 of GBN         046               85.192     248             248   
4  051 NM 022 of GBN         051               94.452     022             022   
5  005 NM 120 of BAL         005                9.260     120             120   
6  010 NM 320 of BRV         010               18.520     320             320   

  IDENT  
0   AML  
1   SXC  
2   SAX  
3   GBN  
4   GBN  
5   BAL  
6   BRV

In [48]:
fr7_nav = link_navaidCleaned(fr_airNONlink)
fr7_nav

index                                            REMARKS       DATASET  \
0    446  Aircraft observed a Red and Black uas at the A...  FIXED RADIAL   
1    449  Aircraft observed a uas off the front while NW...  FIXED RADIAL   
2    450  Aircraft observed a uas 1,500 feet below them ...  FIXED RADIAL   
3    458  Aircraft observed a uas E bound at 16,000 feet...  FIXED RADIAL   
4    459  Aircraft observed a uas while E bound at FL240...  FIXED RADIAL   
5    465  Aircraft observed a white quadcopter uas off t...  FIXED RADIAL   
6    470  Aircraft reported a white quad copter at the B...  FIXED RADIAL   

         UASLOCATION Distance_NM  Distance_Kilometers Bearing Bearing_Degrees  \
0  008 NM 140 of AML         008               14.816     140             140   
1  037 NM 107 of SXC         037               68.524     107             107   
2  015 NM 142 of SAX         015               27.780     142             142   
3  046 NM 248 of GBN         046               85.192     248             248   
4  051 NM 022 of GBN         051               94.452     022             022   
5  005 NM 120 of BAL         005                9.260     120             120   
6  010 NM 320 of BRV         010               18.520     320             320   

  IDENT   LONGITUDE   LATITUDE        NAME_TXT            CITY  
0   AML  -77.466698  38.934604           ARMEL         HERNDON  
1   SXC -118.419922  33.375061  SANTA CATALINA  SANTA CATALINA  
2   SAX  -74.538314  41.067552          SPARTA          SPARTA  
3   GBN -112.674283  32.956263       GILA BEND       GILA BEND  
4   GBN -112.674283  32.956263       GILA BEND       GILA BEND  
5   BAL  -76.661258  39.171071       BALTIMORE       BALTIMORE  
6   BRV  -77.352868  38.336265          BROOKE          BROOKE

In [49]:
fix_radial7 = uas_lat_long(fr7_nav)
fix_radial7

index                                            REMARKS       DATASET  \
0    446  Aircraft observed a Red and Black uas at the A...  FIXED RADIAL   
1    449  Aircraft observed a uas off the front while NW...  FIXED RADIAL   
2    450  Aircraft observed a uas 1,500 feet below them ...  FIXED RADIAL   
3    458  Aircraft observed a uas E bound at 16,000 feet...  FIXED RADIAL   
4    459  Aircraft observed a uas while E bound at FL240...  FIXED RADIAL   
5    465  Aircraft observed a white quadcopter uas off t...  FIXED RADIAL   
6    470  Aircraft reported a white quad copter at the B...  FIXED RADIAL   

         UASLOCATION Distance_NM  Distance_Kilometers Bearing Bearing_Degrees  \
0  008 NM 140 of AML         008               14.816     140             140   
1  037 NM 107 of SXC         037               68.524     107             107   
2  015 NM 142 of SAX         015               27.780     142             142   
3  046 NM 248 of GBN         046               85.192     248             248   
4  051 NM 022 of GBN         051               94.452     022             022   
5  005 NM 120 of BAL         005                9.260     120             120   
6  010 NM 320 of BRV         010               18.520     320             320   

  IDENT   LONGITUDE   LATITUDE        NAME_TXT            CITY  UAS_Latitude  \
0   AML  -77.466698  38.934604           ARMEL         HERNDON     38.832315   
1   SXC -118.419922  33.375061  SANTA CATALINA  SANTA CATALINA     33.192432   
2   SAX  -74.538314  41.067552          SPARTA          SPARTA     40.870252   
3   GBN -112.674283  32.956263       GILA BEND       GILA BEND     32.665658   
4   GBN -112.674283  32.956263       GILA BEND       GILA BEND     33.745274   
5   BAL  -76.661258  39.171071       BALTIMORE       BALTIMORE     39.129330   
6   BRV  -77.352868  38.336265          BROOKE          BROOKE     38.463993   

   UAS_Longitude  
0     -77.357018  
1    -117.717184  
2     -74.335431  
3    -113.516348  
4    -112.292441  
5     -76.568515  
6     -77.489267

In [50]:
# Export to .csv 
fix_radial7.to_csv('fixRadial7_uasLatLong.csv', index = False)